src/blog/first-100k-lines.vmd

```diff
@@ -0,0 +1,415 @@
+
+"That's odd," I said, "This really should be simpler." and I started refactoring.
+
+However, I quickly ran into a wall, and discovered why our code looked odd: it was actually handling some pretty complicated requirements that required that odd approach.
+
+
+In 2021, I came across the same code.
+
+"That's odd," I said, "This really should be simpler." and I started refactoring.
+
+...and proceeded to run into _that same wall._ It was only then that I remembered my first attempt.
+
+
+*This is why we leave comments!*
+
+
+However, comments can become out-of-date, and you might not be lucky enough to stumble across the right comment before you embark on a refactoring adventure.
+
+
+For this reason, we also scatter *clues* around, like the below `see SAIRFU`.
+
+```
+// Here we used to check that the rule's runes were solved, but
+// we don't do that anymore because some rules leave their runes
+// as mysteries, see SAIRFU.
+```
+
+
+
+"See (acronym)" means to search our internal docs for more explanations and discussion. In this case, the documentation's SAIRFU section says:
+
+```
+# Send and Impl Rules For Upcasts (SAIRFU)
+
+`a MyInterface = MyStruct(1337);`
+
+Let's say a's coord rune is X. We can't say:
+`X = MyInterface`
+`X = MyStruct`
+because that's a conflict.
+
+We could just not feed the X = MyStruct into the solver, but then this breaks:
+`a = MyStruct(1337);`
+because it can't figure out what X should be.
+
+...
+```
+
+
+
+This is good for many reasons:
+
+ * We can paste "Some rules leave their runes as mysteries, see SAIRFU." into many comments across our codebase, anywhere that's relevant. This is much better than copying an entire lengthy explanation to a dozen places in the code.
+ * If we later decide SAIRFU wasn't wise, we can search the codebase for SAIRFU to find all of the places we need to update.
+
+
+*TL;DR:* Have many links to centralized documentation.
+
+
+
+# Avoid Nondeterminism
+
+Have you ever had a bug that you could only reproduce one out of every ten tries? What an adventure!
+
+And then there's the bugs that you might only reproduce every hundredth try... *a harrowing endeavor.* You try and you try, and eventually give up and close the bug report as "Cannot Reproduce".
+
+
+This unpredictability is caused by non-determinism, in other words, *when there are some random factors that make it hard to predict whether something will happen.* For example:
+
+ * You're using a random number generator, and the bug only happens when the random number is odd.
+ * The bug only happens when thread A happens to access data before thread B.
+ * The bug only happens when network response B comes before network response A.
+ * You're doing something if C#'s string.GetHashCode is odd. [# I was shocked when string.GetHashCode broke my determinism in the 2020 7DRL Challenge.]
```

Comment:
Link other post

Reply:
Alas, there is no post for that particular year =\